src/appmixer/kafka/connections.js

```diff
@@ -0,0 +1,180 @@
+'use strict';
+
+const RegexParser = require('regex-parser');
+const { kafka } = require('./common.js');
+
+let openConnections = {};
+
+const initializeKafkaConsumer = ({ groupId, authDetails }) => {
+
+    const kafkaMaster = kafka();
+    kafkaMaster.init(authDetails);
+    return kafkaMaster.createConsumer({ groupId });
+};
+
+const initializeKafkaProducer = (authDetails) => {
+
+    const kafkaMaster = kafka();
+    kafkaMaster.init(authDetails);
+    return kafkaMaster.createProducer();
+};
+
+const processMessageHeaders = (headers) => {
+
+    const processedHeaders = {};
+    if (headers) {
+        Object.entries(headers).forEach(([key, value]) => {
+            processedHeaders[key] = Buffer.isBuffer(value) ? value.toString('utf8') : value || '';
+        });
+    }
+    return processedHeaders;
+};
+
+const processMessageData = (message) => {
+
+    return {
+        key: message.key.toString(),
+        value: message.value.toString(),
+        headers: processMessageHeaders(message.headers)
+    };
+};
+
+const addConnection = async (context, component, mode) => {
+
+    const { topics, flowId, componentId, fromBeginning, authDetails, groupId } = component;
+    authDetails.connectionTimeout = context.config.connectionTimeout;
+
+    const connectionId = `${flowId}:${componentId}`;
+    if (openConnections[connectionId]) return; // Connection already exists, do nothing
+
+    await context.service.stateSet(connectionId, { ...component, mode });
```

Comment:
Because there are no locks or anything, I think the following can happen:

- A -> Each -> kafka.SendMessage
- the Each will generate 200 messages, for example
- you have a cluster of 1 node, for simplicity
- roughly 100 messages for kafka.SendMessage will be consumed at the same time (no quota on that component)
- they all will call the plugin API 
- and they all can get to this same line `await context.service.stateSet`
- the result will be, that you will open 100 connections to Kafka. Not one.

Reply:
but @martin-krcmar If there is an open connection, then it simply returns